fixes #4439 - ensure user logins are handled case insensitively

stbenjam opened the pull request.

stbenjam committed:
fixes #4439 - ensure user logins are handled case insensitively

bkearney commented:
Would it be better to improve the validation, and reject? Forcing the change seems nicer to the user, but could be weird for scripts.

stbenjam commented:
We do a lot of fixing of user input elsewhere (hostnames and mac addresses to name 2).

I could change the validation,  but [:alphanum:] also covers foreign alphabets.  Is there a character class that covers that but lowercase only?

The reason I mention that is because today, ستيفن is a perfectly valid username.  Non-latin alphabet users could be broken by this change.  Maybe I'm thinking too much of a fringe case and no one will complain.

bkearney commented:
I was thinking replace [:alnum:] with [:lower:][:digit:]

stbenjam commented:
It breaks a few things.

If there's any users on the planet with non-latin character usernames in Foreman, they'll have to change their username to latin alphabet in order to edit it.  It's probably not a huge deal, but in 2014, I don't like forcing users to use the latin alphabet.

External auth source may be case aware but case-insensitive (Active Directory -- which users are definitely using).  We'll need to have the automatic downcasing anyway, just somewhere else in the code base.

Nothing that can't be overcome, but I don't know if it's worth it.

domcleal commented:
Should we follow AD's lead and be case-aware at least?

stbenjam commented:
@domcleal Ah, I think that's a great compromise.  How's it look? I modified find_by_login from Redmine.  

(For some reason, if doing User.unscoped  then you don't have access to an overridden method, it was using the default rails find_by_login.  That's why I moved unscoped in the method, which I suppose is also DRY anyways).

domcleal commented on app/models/user.rb:
```diff
@@ -175,6 +175,18 @@ def self.try_to_login(login, password)
+      # First look for an exact match
+      user = all(:conditions => {:login => login}).detect {|u| u.login == login}
+      unless user
+        # Fail over to case-insensitive if none was found
```
Can we not just use this search all the time without doing the above search, followed by a second?  The only downside I can think of is if the login field is indexed, the `LOWER(login)` probably won't take advantage.

This would enable this find_\* method to be rewritten as a scope, which means we can use it more easily, and it wouldn't force the user to use User.unscoped (which may not always be appropriate).

stbenjam replied:
Could do, this is just the method that comes from Redmine's User model.  Also all uses of find_by_login in the code were unscoped already.

domcleal replied:
Ok, I think it might make some sense.. for a) performance, and b) if LOWER(login) didn't equal login.downcase, as they're two different implementations.  If you can revert the unscoped changes, I think this will be ok.

stbenjam replied:
Reverted.

domcleal commented:
@adelton is this approach going to cause any issues with REMOTE_USER based authentication?  It works for LDAP and internal auth as far as I can tell.

adelton commented:
Will the change make sure that User.create! creates the lowercased record?

domcleal commented:
@adelton no, the title (and commit message @stbenjam) are inaccurate now.  The PR now makes Foreman do case-insensitive lookups for users.

domcleal commented on db/migrate/20140312224442_add_index_to_lower_login.rb:
```diff
@@ -0,0 +1,13 @@
+
+  def down
+    if ActiveRecord::Base.configurations[Rails.env]['adapter'] == 'postgresql'
+      execute "DROP INDEX index_users_on_lower_loigin"
```
typo, "login"

adelton commented:
Re " MySQL has case-insensitive indices" in https://github.com/stbenjam/foreman/commit/1e61fff45e491c4543b103ddb0f0177b40e93cca#commitcomment-5660692 -- that is not true in all cases. The behaviour of indexes depends on the character set and collation used, and maybe the backend as well.

stbenjam commented:
@adelton Does MySQL support function indexes? AFAIK you have to create a new column, triggers, etc. I would guess it's OK without the MySQL index, unless someone complains about performance...

domcleal commented:
Yeah, I think leaving it without an index is acceptable.

domcleal commented on app/models/user.rb:
```diff
@@ -211,6 +211,16 @@ def self.find_or_create_external_user(attrs, auth_source_name)
 
+  def self.find_by_login(login)
+    # First look for an exact match
+    user = all(:conditions => {:login => login}).detect {|u| u.login == login}
```
Rails 3 style is `where(:login => login)`

domcleal commented:
Sorry for the delay, I've been thinking about this and I think an additional column (which we update with a before_save or similar) actually has a use beyond performance.  Rails uniqueness validators (which we rely on here with case_sensitive = false) are fallible, so a unique index against another column would prevent duplicates on SQLite or MySQL.

stbenjam commented:
Sorry it took so long to get back to this.  I implemented as described.  Since the current status of a database could be that there's multiple users...e.g. stbenjam and STBENJAM, I delete the duplicates in the migration, but I'm not 100% sure which to delete.

My initial inclination was `User.unscoped.all(:order => "created_at ASC")` so I delete all but the oldest.  But the original behavior would be that the only working user would be the first in the default scope, so I should just do User.all and delete any but the first in that order?

ohadlevy commented:
Not sure if it's better but we store the last login date

adelton commented:
> I delete the duplicates in the migration, but I'm not 100% sure which to delete.

The only feasible possibility is not to delete but rename. Both users might have used the application actively and created bunch of database records that you will likely lose. Whichever user you pick to delete, there will be cases when you've picked the wrong one.

Why not rename (put timestamp or something to the login) and create action item or whatever Foreman has as notification queue for admin to sort out the situation and rename / delete manually.

stbenjam commented:
Ok, this renames now and goes through the user list by most recent login first.

I guess a mention in the "upgrade notes" is probably sufficient? The notices model doesn't actually work. This is really a fringe case anyway... duplicate users were never reachable, you were just allowed to create them.

domcleal commented on db/migrate/20140325093623_add_lowerlogin_to_users.rb:
```diff
@@ -0,0 +1,28 @@
+    User.unscoped.order("last_login_on DESC").each do |u|
+      begin
+        u.lower_login = u.login.downcase
+        u.save!
```
This might need to skip validation, as the User model permits an empty e-mail address when creating users, but requires it on update.  My migration exploded as I have users with blank e-mail addresses, and saving failed.

Edit: probably best just to use update_attribute.

stbenjam replied:
@domcleal Ah, Ok.   I didn't use update_attribute originally because if there's a duplicate login, the exception returned is DB-specific and not predictable like RecordInvalid.  Naked exception catching and all.

I can probably refactor it to look for duplicates first, though, and then use update_attribute.

domcleal replied:
Oh right, of course.  Hrmph.

domcleal commented on test/unit/user_test.rb:
```diff
@@ -29,6 +29,26 @@ def setup
+  test "user login should be case aware" do
+    user = User.new :auth_source => auth_sources(:one), :login => "User", :mail  => "foo1@bar.com", :password => "foo"
+    assert user.save
+    assert_equal user.login, "User"
```
Could you assert user.lower_login too please?

stbenjam replied:
Done

dLobatog commented:
Ping @stbenjam

csschwe commented:
Probably need to make sure User IDs pulled from External groups are validated against internal IDs correctly.  I have an issue currently because my external IDs are uppercase in ldap but foreman has them stored as lowercase.  The building of the external groups doesn't work currently because of this mismatch.  It thinks all of the UPPER case IDs need to be added and will remove any LOWER case IDs from the User Group.

stbenjam commented:
@csschwe Ah, interesting - I haven't prioritized this because it's been tricky, and was only theoretical (no one's come complaining about it yet).  

@eLobato knows the user group code better, but I think this PR once cleaned up will fix it if we're correctly storing the case in the database (case-aware but case-insensitive)

dLobatog commented:
That sounds right, this PR should fix the problem of storing lowercase logins from external user groups since now we're case-aware. 
However, there might be the case of a external user that shares its login (but with different casing) with an internal user, how do we solve this problem? With this PR the external user won't be able to log in because login will validate, but lower_login won't and an account won't be created. 
Could you write the code to handle that here?

stbenjam commented:
No, I don't think so, usernames need to be unique on lower_login.  If I understnad your request right, I'd have STBENJAM in LDAP, and an stbenjam local user and I should be able to login with either?

I don't think we should support that for simplicty sake.  Then it's case-aware-but-case-insensitive-(but only-sometimes).

dLobatog commented:
Sorry for being unclear, what I meant was that we should at least display something useful for the user when they attempt that, something like "Your user name in an external authentication source conflicts with an already registered name in Foreman. Please contact your Foreman administrator to solve this.". I don't think we should support logging in in that case either.

dLobatog commented:
@stbenjam Hey any updates on this? It seems like the code was in a fairly good state, ping me for review after a rebase and a fix for that test @domcleal asked for.

stbenjam commented:
@eLobato Pong! Sorry for the delay :-)

Updated.

dLobatog commented:
Merged as 6be0508bfb1745c4aa4b934744b732a520502721 thanks @stbenjam !
